Bug 1518261

k88hudson opened the pull request:
This patch adds a periodic check after every system tick (i.e. 5 minutes) and refactors some of the fetching into a single function. If the cache has expired, a refresh will occur.

I think this one is going to be hard to manually QA, but I'll try to think of something

ScottDowne commented on lib/DiscoveryStreamFeed.jsm:
```diff
@@ -67,54 +67,31 @@ this.DiscoveryStreamFeed = class DiscoveryStreamFeed {
+  async fetchFromEndpoint(endpoint) {
     if (!endpoint) {
-      Cu.reportError("No endpoint configured for pocket, so could not fetch layout");
+      Cu.reportError("Tried to fetch endpoint but none was configured.");
```
How do errors get reported? Is it obvious where the error is coming from, or would it be helpful to add a bit of context to the error, example "Tried to fetch Discovery Stream endpoint but none was configured."

k88hudson replied:
You do get a line number / stack trace, so that should be enough (although I could improve this by maybe adding a reference to the endpoint?

ScottDowne replied:
Line numbers is fine, let's not worry about this too much then.

ScottDowne commented on lib/DiscoveryStreamFeed.jsm:
```diff
@@ -279,6 +250,12 @@ this.DiscoveryStreamFeed = class DiscoveryStreamFeed {
+      case at.SYSTEM_TICK:
+        // Only refresh if we loaded once in .enable()
+        if (this.config.enabled && this.loaded) {
+          await this.refreshAll({updateOpenTabs: false});
```
I think this'll end up firing update events on every system tick.

If I'm reading it correctly, it goes into a layout/component/spoc load function, gets cache. If it needs to get fresh data, it does. Then it sends the fresh data or cache to the reducer.

The old code was slightly less dry, but also only fired system tick updates if cache has expired.

k88hudson commented:
Ok, I have added some more logic that covers checking cache expiry – I think ideally it would be nicer if we added that right to the cache API itself.
